packages/editor/src/optimizer/common.ts

```diff
@@ -0,0 +1,614 @@
+  tag: T;
+  variant: "Ok";
+  value: V;
+};
+
+export type TaggedErr<E, T> = {
+  tag: T;
+  variant: "Err";
+  error: E;
+};
+
+/**
+ * A Result type, but with a tag. Supertype of `MessageResult`.
+ */
+export type TaggedResult<V, E, T> = TaggedOk<V, T> | TaggedErr<E, T>;
+
+export type CompileResult = TaggedResult<
+  CompileInfo,
+  PenroseError,
+  MessageTags.Compile
+>;
+
+export type PollResult = TaggedResult<
+  HistoryInfo,
+  InvalidDiagramIDError,
+  MessageTags.Poll
+>;
+
+export type ComputeLayoutResult = TaggedResult<
+  LayoutState,
+  InvalidDiagramIDError | InvalidHistoryLocError | PenroseError,
+  MessageTags.ComputeLayout
+>;
+
+export type ResampleResult = TaggedResult<
+  StepSequenceID,
+  InvalidDiagramIDError,
+  MessageTags.Resample
+>;
+
+// Errors
+
+export type InvalidDiagramIDError = {
+  tag: "InvalidDiagramIDError";
+  id: DiagramID;
+  validIds: Set<DiagramID>;
+};
+
+export type InvalidHistoryLocError = {
+  tag: "InvalidHistoryLocError";
+  historyLoc: HistoryLoc;
+  historyInfo: HistoryInfo;
+};
+export type OptimizationError = {
+  tag: "OptimizationError";
+  error: unknown;
+};
+
+export type OptimizerError =
+  | OptimizationError
+  | InvalidDiagramIDError
+  | InvalidHistoryLocError;
+
+// Utils
+
+class NumericIDGenerator {
+  private nextId = 0;
+
+  next = () => {
+    return this.nextId++;
+  };
+}
+
+// We reexport as different generators for each ID type, in case we decide
+// not to use numbers in the future
+export {
+  NumericIDGenerator as DiagramIDGenerator,
+  NumericIDGenerator as MessageIDGenerator,
+  NumericIDGenerator as StepSequenceIDGenerator,
+};
+
+export type Tagged<T> = {
+  tag: T;
+};
+
+/**
+ * Make a request to a worker.
+ *
+ * @param worker Worker to request
+ * @param data Request data
+ * @param resolvesById Map from `MessageID` to `MessageResolve`. Mutated.
+ * @param messageIdGenerator Generator from which we should get the next message id
+ */
+export const request = <T, R extends MessageResult & Tagged<T>>(
+  worker: Worker,
+  data: MessageRequestData & Tagged<T>,
+  resolvesById: Map<MessageID, MessageResolve>,
+  messageIdGenerator: NumericIDGenerator,
+): Promise<R> => {
+  return new Promise((resolve) => {
+    const messageId: MessageID = messageIdGenerator.next();
+    const request: MessageRequest = {
+      tag: "MessageRequest",
+      messageId,
+      data,
+    };
+    resolvesById.set(messageId, (result: MessageResult) => {
+      if (result.tag === data.tag) {
+        (resolve as any)(result);
+        resolvesById.delete(messageId);
+      } else {
+        throw new Error(
+          `MessageID ${messageId} expected result of type 
+          ${data.tag} but received ${result.tag}`,
+        );
+      }
+    });
+    worker.postMessage(request);
+  });
+};
+
+/**
+ * Launch worker, and wait for a notification `InitData` before resolving.
+ * @param url Worker to spin
+ */
+export const spinAndWaitForInit = async (url: string): Promise<Worker> => {
```

Comment:
When I hear "spin" I think that the thread is blocking, running an infinite `for (;;)` loop or something. Just to clarify, that's not happening here, right? In that case, can you change the name so it doesn't contain the word "spin"?

Reply:
Oh I think I understand, does "spin" mean "spin up"? I think it's fine to say "spin up" then, just not "spin" by itself.

Reply:
I think launch is maybe just a better word.

Reply:
was also expecting a `while(true)` before I jumped to definition. `waitForInit` might just be fine?